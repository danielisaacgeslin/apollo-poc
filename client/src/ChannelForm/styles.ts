import { css } from 'emotion';

export const container = css`
  padding: 15px 0;
  max-width: 300px;
`;

export const buttonContainer = css`
  margin-top: 15px;
`;

export const controlContainer = css`
  display: flex;
  flex-direction: column;
`;

export const input = css`
  margin-bottom: 10px;
`;
