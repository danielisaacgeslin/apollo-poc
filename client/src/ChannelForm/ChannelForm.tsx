import React, { memo, useState, useCallback, ChangeEvent, useEffect } from 'react';

import { SuccessButton } from '../shared/SuccessButton';
import { Button } from '../shared/Button';
import { Select } from '../shared/Select';
import { Input } from '../shared/Input';
import { buttonContainer, container, controlContainer, input } from './styles';

const types = ['dm', 'group'];
const initialState = { name: '', type: types[0] };

export interface Props {
  isLoading: boolean;
  onSave: (name: string, type: string) => void;
}

const ChannelForm = ({ isLoading, onSave }: Props) => {
  const [state, setState] = useState<{ name: string; type: string }>(initialState);

  const onNameChange = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    const { value: name } = event.target;
    setState(prevState => ({ ...prevState, name }));
  }, []);
  const onTypeChange = useCallback((event: ChangeEvent<HTMLSelectElement>) => {
    const { value: type } = event.target;
    setState(prevState => ({ ...prevState, type }));
  }, []);
  const onCancel = useCallback(() => setState(prevState => ({ ...prevState, ...initialState })), []);
  const onSaveChannel = useCallback(() => onSave(state.name, state.type), [state, onSave]);

  useEffect(() => onCancel(), [isLoading, onCancel]);

  return (
    <div className={container}>
      <h4>create a new channel</h4>
      <div className={controlContainer}>
        <Input placeholder="type a channel name" value={state.name} onChange={onNameChange} className={input} />
        <Select value={state.type} onChange={onTypeChange}>
          {types.map(option => (
            <option key={option} value={option}>
              {option}
            </option>
          ))}
        </Select>
      </div>
      <div className={buttonContainer}>
        <SuccessButton disabled={isLoading || !state.name} style={{ marginRight: 5 }} onClick={onSaveChannel}>
          save
        </SuccessButton>
        <Button disabled={isLoading} onClick={onCancel}>
          cancel
        </Button>
      </div>
    </div>
  );
};

export default memo(ChannelForm);
