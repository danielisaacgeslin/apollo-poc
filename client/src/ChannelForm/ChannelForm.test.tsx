import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { default as ChannelForm, Props, types } from './ChannelForm';

describe('ChannelForm', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      isLoading: false,
      onSave: jest.fn()
    };
  });

  it('should render', () => {
    const { container } = render(<ChannelForm {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render loading', () => {
    const { container } = render(<ChannelForm {...props} isLoading={true} />);
    expect(container).toMatchSnapshot();
  });

  it('should create a channel', () => {
    const name = 'a channel';
    const [_, type] = types;
    const { getByTestId } = render(<ChannelForm {...props} />);
    fireEvent.change(getByTestId('channel-name'), { target: { value: name } });
    fireEvent.change(getByTestId('channel-type'), { target: { value: type } });
    fireEvent.click(getByTestId('channel-save'));
    expect(props.onSave).toBeCalledWith(name, type);
  });

  it('should NOT create while loading', () => {
    const name = 'a channel';
    const { getByTestId } = render(<ChannelForm {...props} isLoading={true} />);
    fireEvent.change(getByTestId('channel-name'), { target: { value: name } });
    fireEvent.click(getByTestId('channel-save'));
    expect(props.onSave).not.toBeCalled();
  });

  it('should NOT create with an empty name', () => {
    const { getByTestId } = render(<ChannelForm {...props} isLoading={true} />);
    fireEvent.click(getByTestId('channel-save'));
    expect(props.onSave).not.toBeCalled();
  });

  it('should cancel', () => {
    const name = 'a channel';
    const { getByTestId } = render(<ChannelForm {...props} />);
    fireEvent.change(getByTestId('channel-name'), { target: { value: name } });
    expect((getByTestId('channel-name') as HTMLInputElement).value).toEqual(name);
    fireEvent.click(getByTestId('channel-cancel'));
    expect((getByTestId('channel-name') as HTMLInputElement).value).toEqual('');
  });
});
