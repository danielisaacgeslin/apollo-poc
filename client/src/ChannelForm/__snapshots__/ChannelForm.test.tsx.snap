// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelForm should render 1`] = `
.emotion-7 {
  padding: 15px 0;
  max-width: 300px;
}

.emotion-6 {
  margin-top: 15px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  margin-bottom: 10px;
}

.emotion-1 {
  height: 30px;
  padding: 2px 5px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-2 {
  height: 30px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-4 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  background-color: #3AC6AB;
  color: #FFFFFF;
}

.emotion-5 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

<div>
  <div
    class="emotion-7"
  >
    <h4>
      create a new channel
    </h4>
    <div
      class="emotion-3"
    >
      <input
        class="emotion-0 emotion-1"
        data-testid="channel-name"
        placeholder="type a channel name"
        value=""
      />
      <select
        class="emotion-2"
        data-testid="channel-type"
      >
        <option
          value="dm"
        >
          dm
        </option>
        <option
          value="group"
        >
          group
        </option>
      </select>
    </div>
    <div
      class="emotion-6"
    >
      <button
        class="emotion-4"
        data-testid="channel-save"
        disabled=""
        style="margin-right: 5px;"
      >
        save
      </button>
      <button
        class="emotion-5"
        data-testid="channel-cancel"
      >
        cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`ChannelForm should render loading 1`] = `
.emotion-7 {
  padding: 15px 0;
  max-width: 300px;
}

.emotion-6 {
  margin-top: 15px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  margin-bottom: 10px;
}

.emotion-1 {
  height: 30px;
  padding: 2px 5px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-2 {
  height: 30px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-4 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  background-color: #3AC6AB;
  color: #FFFFFF;
}

.emotion-5 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

<div>
  <div
    class="emotion-7"
  >
    <h4>
      create a new channel
    </h4>
    <div
      class="emotion-3"
    >
      <input
        class="emotion-0 emotion-1"
        data-testid="channel-name"
        placeholder="type a channel name"
        value=""
      />
      <select
        class="emotion-2"
        data-testid="channel-type"
      >
        <option
          value="dm"
        >
          dm
        </option>
        <option
          value="group"
        >
          group
        </option>
      </select>
    </div>
    <div
      class="emotion-6"
    >
      <button
        class="emotion-4"
        data-testid="channel-save"
        disabled=""
        style="margin-right: 5px;"
      >
        save
      </button>
      <button
        class="emotion-5"
        data-testid="channel-cancel"
        disabled=""
      >
        cancel
      </button>
    </div>
  </div>
</div>
`;
